File: python/8.py
#!/usr/bin/python3
import queue

N = 3
SIZE = N * N

positions = {}

def getKey(board):
    r = 0
    for x in board:
        r = r * 9 + x
    return r

def findEmpty(board):
    for i in range(SIZE):
        if board[i] == 0:
            return i
    return -1

def printSolution(board):
    dir = positions[getKey(board)]
    empty = findEmpty(board)
    next = -1
    if dir == 1:
        next = empty + 1
    elif dir == 2:
        next = empty - 1
    elif dir == 3:
        next = empty + N
    elif dir == 4:
        next = empty - N
    if next != -1:
        board[next], board[empty] = board[empty], board[next]
        printSolution(board)
        board[next], board[empty] = board[empty], board[next]

    for x in board:
        print(x, end='')
    print()


start = [int(x) for x in input().split()]
b = [0] * SIZE
for s in start:
    if s >= 0 and s < SIZE:
        b[s] = 1
# проверить ввод
if sum(b) != SIZE:
    print('Некорректная позиция')
    exit(0)

q = queue.Queue()
q.put(start)
positions[getKey(start)] = 0
while not q.empty():
    board = q.get()
    empty = findEmpty(board)
    # перебираем все направления движения
    for i in range(1, 5):
        next = -1
        if i == 1 and empty % N != 0:
            next = empty - 1
        elif i == 2 and empty % N != N - 1:
            next = empty + 1
        elif i == 3 and empty >= N:
            next = empty - N
        elif i == 4 and empty < N * N - N:
            next = empty + N
        if next != -1:
            b = board.copy()
            b[empty], b[next] = b[next], b[empty]
            key = getKey(b)
            if key not in positions:
                positions[key] = i
                q.put(b)

final = [1, 2, 3, 4, 5, 6, 7, 8, 0]
if getKey(final) not in positions:
    print('Решения не существует!')
else:
    printSolution(final)

File: cpp/8.cpp
#include <iostream>
#include <array>
#include <map>
#include <queue>

const int N = 3;
const int SIZE = N * N;

typedef std::array<int, SIZE> Board;

int getKey(const Board &board)
{
    int r = 0;
    for (auto x : board)
    {
        r = r * 9 + x;
    }
    return r;
}

typedef std::map<long long, int> Positions;
Positions positions;

int findEmpty(const Board &board)
{
    for (int i = 0 ; i < SIZE ; ++i)
    {
        if (board[i] == 0)
        {
            return i;
        }
    }
    return -1;
}

void printSolution(Board &board)
{
    int dir = positions[getKey(board)];
    int empty = findEmpty(board);
    int next = -1;
    if (dir == 1)
    {
        next = empty + 1;
    }
    else if (dir == 2)
    {
        next = empty - 1;
    }
    else if (dir == 3)
    {
        next = empty + N;
    }
    else if (dir == 4)
    {
        next = empty - N;
    }
    if (next != -1)
    {
        std::swap(board[next], board[empty]);
        printSolution(board);
        std::swap(board[next], board[empty]);
    }
    for (auto x : board)
    {
        std::cout << x;
    }
    std::cout << "\n";
}

int main()
{
    Board start;
    bool b[SIZE] = {};
    for (int i = 0 ; i < SIZE ; ++i)
    {
        std::cin >> start[i];
        if (start[i] >= 0 && start[i] < SIZE)
        {
            b[start[i]] = true;
        }
    }
    // проверить ввод
    for (int i = 0 ; i < SIZE ; ++i)
    {
        if (!b[i])
        {
            std::cout << "Некорректная позиция\n";
            return 0;
        }
    }

    std::queue<Board> q;
    q.push(start);
    positions[getKey(start)] = -1;
    while (!q.empty())
    {
        Board board = q.front();
        q.pop();
        int empty = findEmpty(board);
        // перебираем все направления движения
        for (int i = 1 ; i <= 4 ; ++i)
        {
            int next = -1;
            if (i == 1 && empty % N != 0)
            {
                next = empty - 1;
            }
            else if (i == 2 && empty % N != N - 1)
            {
                next = empty + 1;
            }
            else if (i == 3 && empty >= N)
            {
                next = empty - N;
            }
            else if (i == 4 && empty < N * N - N)
            {
                next = empty + N;
            }
            if (next != -1)
            {
                Board b = board;
                std::swap(b[empty], b[next]);
                int key = getKey(b);
                if (!positions.contains(key))
                {
                    positions[key] = i;
                    q.push(b);
                }
            }
        }
    }

    Board final = {1, 2, 3, 4, 5, 6, 7, 8, 0};
    if (!positions.contains(getKey(final)))
    {
        std::cout << "Решения не существует!\n";
    }
    else
    {
        printSolution(final);
    }
}
